File: src/data/langchain-integration-for-sap-hana-cloud-main/mcp_server/kuzu_store.py
# SPDX-License-Identifier: Apache-2.0
# SPDX-FileCopyrightText: 2024 SAP SE
"""
KùzuDB Graph-RAG store for langchain-integration-for-sap-hana-cloud MCP server.

Schema:
  Nodes  : HanaVectorStore, EmbeddingDeployment, HanaSchema
  Edges  : USES_DEPLOYMENT  (HanaVectorStore → EmbeddingDeployment)
           LIVES_IN          (HanaVectorStore → HanaSchema)
           RELATED_SCHEMA    (HanaSchema → HanaSchema)

Gracefully degrades when the `kuzu` Python package is not installed.
"""
from __future__ import annotations

import os
from typing import Any

# ---------------------------------------------------------------------------
# KuzuStore
# ---------------------------------------------------------------------------

class KuzuStore:
    """Embedded KùzuDB graph store for HANA LangChain entity relationships."""

    def __init__(self, db_path: str = ".kuzu-langchain-hana") -> None:
        self._db_path = db_path
        self._db: Any = None
        self._conn: Any = None
        self._available: bool = False
        self._schema_ready: bool = False
        self._init()

    def _init(self) -> None:
        try:
            import kuzu  # type: ignore[import]
            self._db = kuzu.Database(self._db_path)
            self._conn = kuzu.Connection(self._db)
            self._available = True
        except Exception:
            self._available = False

    def available(self) -> bool:
        return self._available

    # ------------------------------------------------------------------
    # Schema
    # ------------------------------------------------------------------

    def ensure_schema(self) -> None:
        if not self._available or not self._conn or self._schema_ready:
            return

        ddl = [
            # Node tables
            """CREATE NODE TABLE IF NOT EXISTS HanaVectorStore (
                storeId        STRING,
                tableName      STRING,
                embeddingModel STRING,
                schema         STRING,
                PRIMARY KEY (storeId)
            )""",
            """CREATE NODE TABLE IF NOT EXISTS EmbeddingDeployment (
                deploymentId  STRING,
                modelName     STRING,
                resourceGroup STRING,
                status        STRING,
                PRIMARY KEY (deploymentId)
            )""",
            """CREATE NODE TABLE IF NOT EXISTS HanaSchema (
                schemaId       STRING,
                schemaName     STRING,
                classification STRING,
                PRIMARY KEY (schemaId)
            )""",
            # Relationship tables
            """CREATE REL TABLE IF NOT EXISTS USES_DEPLOYMENT (
                FROM HanaVectorStore TO EmbeddingDeployment
            )""",
            """CREATE REL TABLE IF NOT EXISTS LIVES_IN (
                FROM HanaVectorStore TO HanaSchema
            )""",
            """CREATE REL TABLE IF NOT EXISTS RELATED_SCHEMA (
                FROM HanaSchema TO HanaSchema
            )""",
        ]

        for stmt in ddl:
            self._conn.execute(stmt)

        self._schema_ready = True

    # ------------------------------------------------------------------
    # Upsert helpers
    # ------------------------------------------------------------------

    def upsert_vector_store(
        self,
        store_id: str,
        table_name: str,
        embedding_model: str,
        schema: str,
    ) -> None:
        if not self._available or not self._conn:
            return
        self._conn.execute(
            "MERGE (s:HanaVectorStore {storeId: $id}) "
            "ON MATCH SET s.tableName = $tn, s.embeddingModel = $em, s.schema = $sc "
            "ON CREATE SET s.tableName = $tn, s.embeddingModel = $em, s.schema = $sc",
            {"id": store_id, "tn": table_name, "em": embedding_model, "sc": schema},
        )

    def upsert_deployment(
        self,
        deployment_id: str,
        model_name: str,
        resource_group: str,
        status: str,
    ) -> None:
        if not self._available or not self._conn:
            return
        self._conn.execute(
            "MERGE (d:EmbeddingDeployment {deploymentId: $id}) "
            "ON MATCH SET d.modelName = $mn, d.resourceGroup = $rg, d.status = $st "
            "ON CREATE SET d.modelName = $mn, d.resourceGroup = $rg, d.status = $st",
            {"id": deployment_id, "mn": model_name, "rg": resource_group, "st": status},
        )

    def upsert_schema(
        self,
        schema_id: str,
        schema_name: str,
        classification: str,
    ) -> None:
        if not self._available or not self._conn:
            return
        self._conn.execute(
            "MERGE (h:HanaSchema {schemaId: $id}) "
            "ON MATCH SET h.schemaName = $sn, h.classification = $cl "
            "ON CREATE SET h.schemaName = $sn, h.classification = $cl",
            {"id": schema_id, "sn": schema_name, "cl": classification},
        )

    # ------------------------------------------------------------------
    # Link helpers
    # ------------------------------------------------------------------

    def link_store_deployment(self, store_id: str, deployment_id: str) -> None:
        if not self._available or not self._conn:
            return
        self._conn.execute(
            "MATCH (s:HanaVectorStore {storeId: $sid}), "
            "(d:EmbeddingDeployment {deploymentId: $did}) "
            "MERGE (s)-[:USES_DEPLOYMENT]->(d)",
            {"sid": store_id, "did": deployment_id},
        )

    def link_store_schema(self, store_id: str, schema_id: str) -> None:
        if not self._available or not self._conn:
            return
        self._conn.execute(
            "MATCH (s:HanaVectorStore {storeId: $sid}), "
            "(h:HanaSchema {schemaId: $hid}) "
            "MERGE (s)-[:LIVES_IN]->(h)",
            {"sid": store_id, "hid": schema_id},
        )

    def link_schemas(self, from_id: str, to_id: str) -> None:
        if not self._available or not self._conn:
            return
        self._conn.execute(
            "MATCH (a:HanaSchema {schemaId: $from}), (b:HanaSchema {schemaId: $to}) "
            "MERGE (a)-[:RELATED_SCHEMA]->(b)",
            {"from": from_id, "to": to_id},
        )

    # ------------------------------------------------------------------
    # Query helpers
    # ------------------------------------------------------------------

    def run_query(self, cypher: str, params: dict | None = None) -> list[dict]:
        if not self._available or not self._conn:
            return []
        try:
            result = self._conn.execute(cypher, params or {})
            cols = result.get_column_names()
            rows: list[dict] = []
            while result.has_next():
                values = result.get_next()
                rows.append(dict(zip(cols, values)))
            return rows
        except Exception as exc:
            print(f"KùzuDB query failed: {exc}")
            return []

    def get_store_context(self, store_id: str) -> list[dict]:
        """Return embedding deployments connected to a vector store."""
        if not self._available:
            return []
        return self.run_query(
            "MATCH (s:HanaVectorStore {storeId: $id})-[:USES_DEPLOYMENT]->(d:EmbeddingDeployment) "
            "RETURN d.deploymentId AS deploymentId, d.modelName AS modelName, "
            "d.resourceGroup AS resourceGroup, d.status AS status, "
            "'uses_deployment' AS relation",
            {"id": store_id},
        )

    def get_schema_context(self, schema_id: str) -> list[dict]:
        """Return vector stores that live in a given HANA schema."""
        if not self._available:
            return []
        return self.run_query(
            "MATCH (s:HanaVectorStore)-[:LIVES_IN]->(h:HanaSchema {schemaId: $id}) "
            "RETURN s.storeId AS storeId, s.tableName AS tableName, "
            "s.embeddingModel AS embeddingModel, 'lives_in' AS relation",
            {"id": schema_id},
        )


# ---------------------------------------------------------------------------
# Singleton
# ---------------------------------------------------------------------------

_instance: KuzuStore | None = None


def get_kuzu_store() -> KuzuStore:
    global _instance
    if _instance is None:
        _instance = KuzuStore(os.environ.get("KUZU_DB_PATH", ".kuzu-langchain-hana"))
    return _instance


def _reset_kuzu_store() -> None:
    global _instance
    _instance = None

File: archive/cap-llm-plugin-main/mcp-server/src/kuzu-store.ts
// SPDX-License-Identifier: Apache-2.0
// SPDX-FileCopyrightText: 2024 SAP SE
/**
 * KùzuDB Graph-RAG store for cap-llm-plugin MCP server.
 *
 * Schema:
 *   Nodes  : CapService, LlmDeployment, RagTable
 *   Edges  : SERVED_BY  (CapService → LlmDeployment)
 *            USES_TABLE (CapService → RagTable)
 *            ROUTES_TO  (CapService → CapService)
 *
 * Gracefully degrades when the `kuzu` npm package is not installed.
 */

export interface GraphContextEntry {
  relation: string;
  [key: string]: unknown;
}

// ---------------------------------------------------------------------------
// Kuzu type shims (resolved at runtime; absent when package is missing)
// ---------------------------------------------------------------------------

interface KuzuDatabase {
  constructor: new (path: string) => KuzuDatabase;
}

interface KuzuQueryResult {
  getColumnNames(): string[];
  hasNext(): boolean;
  getNext(): unknown[];
}

interface KuzuConnection {
  execute(cypher: string, params?: Record<string, unknown>): Promise<KuzuQueryResult>;
}

// ---------------------------------------------------------------------------
// KuzuStore
// ---------------------------------------------------------------------------

export class KuzuStore {
  private dbPath: string;
  private db: KuzuDatabase | null = null;
  private conn: KuzuConnection | null = null;
  private _available = false;
  private _schemaReady = false;

  constructor(dbPath = '.kuzu-cap-llm') {
    this.dbPath = dbPath;
    this._init();
  }

  private _init(): void {
    try {
      // eslint-disable-next-line @typescript-eslint/no-require-imports
      const kuzu = require('kuzu') as {
        Database: new (path: string) => KuzuDatabase;
        Connection: new (db: KuzuDatabase) => KuzuConnection;
      };
      this.db = new kuzu.Database(this.dbPath);
      this.conn = new kuzu.Connection(this.db);
      this._available = true;
    } catch {
      this._available = false;
    }
  }

  available(): boolean {
    return this._available;
  }

  async ensureSchema(): Promise<void> {
    if (!this._available || !this.conn || this._schemaReady) return;

    const ddl: string[] = [
      // Node tables
      `CREATE NODE TABLE IF NOT EXISTS CapService (
         serviceId   STRING,
         serviceName STRING,
         serviceType STRING,
         dataClass   STRING,
         PRIMARY KEY (serviceId)
       )`,
      `CREATE NODE TABLE IF NOT EXISTS LlmDeployment (
         deploymentId  STRING,
         modelName     STRING,
         resourceGroup STRING,
         status        STRING,
         PRIMARY KEY (deploymentId)
       )`,
      `CREATE NODE TABLE IF NOT EXISTS RagTable (
         tableId     STRING,
         tableName   STRING,
         description STRING,
         schema      STRING,
         PRIMARY KEY (tableId)
       )`,
      // Relationship tables
      `CREATE REL TABLE IF NOT EXISTS SERVED_BY (
         FROM CapService TO LlmDeployment
       )`,
      `CREATE REL TABLE IF NOT EXISTS USES_TABLE (
         FROM CapService TO RagTable
       )`,
      `CREATE REL TABLE IF NOT EXISTS ROUTES_TO (
         FROM CapService TO CapService
       )`,
    ];

    for (const stmt of ddl) {
      await this.conn.execute(stmt);
    }
    this._schemaReady = true;
  }

  // ---------------------------------------------------------------------------
  // Upsert helpers
  // ---------------------------------------------------------------------------

  async upsertService(
    serviceId: string,
    serviceName: string,
    serviceType: string,
    dataClass: string,
  ): Promise<void> {
    if (!this._available || !this.conn) return;
    await this.conn.execute(
      `MERGE (s:CapService {serviceId: $id})
       ON MATCH SET s.serviceName = $name, s.serviceType = $type, s.dataClass = $dc
       ON CREATE SET s.serviceName = $name, s.serviceType = $type, s.dataClass = $dc`,
      { id: serviceId, name: serviceName, type: serviceType, dc: dataClass },
    );
  }

  async upsertDeployment(
    deploymentId: string,
    modelName: string,
    resourceGroup: string,
    status: string,
  ): Promise<void> {
    if (!this._available || !this.conn) return;
    await this.conn.execute(
      `MERGE (d:LlmDeployment {deploymentId: $id})
       ON MATCH SET d.modelName = $model, d.resourceGroup = $rg, d.status = $status
       ON CREATE SET d.modelName = $model, d.resourceGroup = $rg, d.status = $status`,
      { id: deploymentId, model: modelName, rg: resourceGroup, status },
    );
  }

  async upsertRagTable(
    tableId: string,
    tableName: string,
    description: string,
    schema: string,
  ): Promise<void> {
    if (!this._available || !this.conn) return;
    await this.conn.execute(
      `MERGE (t:RagTable {tableId: $id})
       ON MATCH SET t.tableName = $name, t.description = $desc, t.schema = $schema
       ON CREATE SET t.tableName = $name, t.description = $desc, t.schema = $schema`,
      { id: tableId, name: tableName, desc: description, schema },
    );
  }

  // ---------------------------------------------------------------------------
  // Linking helpers
  // ---------------------------------------------------------------------------

  async linkServiceDeployment(serviceId: string, deploymentId: string): Promise<void> {
    if (!this._available || !this.conn) return;
    await this.conn.execute(
      `MATCH (s:CapService {serviceId: $sid}), (d:LlmDeployment {deploymentId: $did})
       MERGE (s)-[:SERVED_BY]->(d)`,
      { sid: serviceId, did: deploymentId },
    );
  }

  async linkServiceTable(serviceId: string, tableId: string): Promise<void> {
    if (!this._available || !this.conn) return;
    await this.conn.execute(
      `MATCH (s:CapService {serviceId: $sid}), (t:RagTable {tableId: $tid})
       MERGE (s)-[:USES_TABLE]->(t)`,
      { sid: serviceId, tid: tableId },
    );
  }

  async linkServiceRoute(fromServiceId: string, toServiceId: string): Promise<void> {
    if (!this._available || !this.conn) return;
    await this.conn.execute(
      `MATCH (a:CapService {serviceId: $from}), (b:CapService {serviceId: $to})
       MERGE (a)-[:ROUTES_TO]->(b)`,
      { from: fromServiceId, to: toServiceId },
    );
  }

  // ---------------------------------------------------------------------------
  // Query helpers
  // ---------------------------------------------------------------------------

  async runQuery(cypher: string, params: Record<string, unknown> = {}): Promise<Record<string, unknown>[]> {
    if (!this._available || !this.conn) return [];
    try {
      const result = await this.conn.execute(cypher, params);
      const cols = result.getColumnNames();
      const rows: Record<string, unknown>[] = [];
      while (result.hasNext()) {
        const values = result.getNext();
        const row: Record<string, unknown> = {};
        cols.forEach((col, i) => { row[col] = values[i]; });
        rows.push(row);
      }
      return rows;
    } catch (err) {
      console.error('KùzuDB query failed:', (err as Error).message);
      return [];
    }
  }

  async getServiceContext(serviceId: string): Promise<GraphContextEntry[]> {
    if (!this._available) return [];
    const rows = await this.runQuery(
      `MATCH (s:CapService {serviceId: $id})-[:SERVED_BY]->(d:LlmDeployment)
       RETURN d.deploymentId AS deploymentId, d.modelName AS modelName,
              d.resourceGroup AS resourceGroup, d.status AS status,
              'served_by' AS relation`,
      { id: serviceId },
    );
    return rows as GraphContextEntry[];
  }

  async getRagContext(tableId: string): Promise<GraphContextEntry[]> {
    if (!this._available) return [];
    const rows = await this.runQuery(
      `MATCH (s:CapService)-[:USES_TABLE]->(t:RagTable {tableId: $id})
       RETURN s.serviceId AS serviceId, s.serviceName AS serviceName,
              s.serviceType AS serviceType, 'uses_table' AS relation`,
      { id: tableId },
    );
    return rows as GraphContextEntry[];
  }
}

// ---------------------------------------------------------------------------
// Singleton
// ---------------------------------------------------------------------------

let _instance: KuzuStore | null = null;

export function getKuzuStore(): KuzuStore {
  if (!_instance) {
    _instance = new KuzuStore(process.env.KUZU_DB_PATH || '.kuzu-cap-llm');
  }
  return _instance;
}

export function _resetKuzuStore(): void {
  _instance = null;
}
